src/libutils/oglapphelpers/msl.mm

```diff
@@ -0,0 +1,504 @@
+    std::vector<float> float4AdaptedLutValues;
+    if(lutValues != nullptr)
+    {
+        float4AdaptedLutValues.resize(valueCount);
+        const float *rgbLutValuesIt = lutValues;
+        float *rgbaLutValuesIt = float4AdaptedLutValues.data();
+        const float *end = rgbaLutValuesIt + valueCount;
+            
+        while(rgbaLutValuesIt != end) {
+            *rgbaLutValuesIt++ = *rgbLutValuesIt++;
+            *rgbaLutValuesIt++ = *rgbLutValuesIt++;
+            *rgbaLutValuesIt++ = *rgbLutValuesIt++;
+            *rgbaLutValuesIt++ = 1.0f;
+        }
+    }
+    
+    return float4AdaptedLutValues;
+}
+
+namespace
+{
+
+id<MTLSamplerState> GetSamplerState(id<MTLDevice> device, Interpolation interpolation)
+{
+    static std::unordered_map<int, id<MTLSamplerState>> samplerStateCache;
+    
+    if(samplerStateCache.find(interpolation) != samplerStateCache.end())
+        return samplerStateCache[interpolation];
+    
+    MTLSamplerDescriptor* samplerDesc = [[MTLSamplerDescriptor new] autorelease];
+    [samplerDesc setMinFilter:MTLSamplerMinMagFilterNearest];
+    
+    if(interpolation==INTERP_NEAREST)
+    {
+        [samplerDesc setMinFilter:MTLSamplerMinMagFilterNearest];
+        [samplerDesc setMagFilter:MTLSamplerMinMagFilterNearest];
+    }
+    else
+    {
+        [samplerDesc setMinFilter:MTLSamplerMinMagFilterLinear];
+        [samplerDesc setMagFilter:MTLSamplerMinMagFilterLinear];
+    }
+
+    [samplerDesc setSAddressMode:MTLSamplerAddressModeClampToEdge];
+    [samplerDesc setTAddressMode:MTLSamplerAddressModeClampToEdge];
+    [samplerDesc setRAddressMode:MTLSamplerAddressModeClampToEdge];
+    
+    samplerStateCache[interpolation] = [[device newSamplerStateWithDescriptor:samplerDesc] autorelease];
+    
+    return samplerStateCache[interpolation];
+}
+
+id<MTLTexture> AllocateTexture3D(id<MTLDevice> device,
+                       unsigned edgelen, const float * lutValues)
+{
+    if(lutValues==0x0)
+    {
+        throw Exception("Missing texture data");
+    }
+    
+    // MTLPixelFormatRGB32Float not supported on metal. Adapt to MTLPixelFormatRGBA32Float
+    std::vector<float> float4AdaptedLutValues = RGB_to_RGBA(lutValues, 3*edgelen*edgelen*edgelen);
+    
+    MTLTextureDescriptor* texDescriptor = [[MTLTextureDescriptor new] autorelease];
+    
+    [texDescriptor setTextureType:MTLTextureType3D];
+    [texDescriptor setWidth:edgelen];
+    [texDescriptor setHeight:edgelen];
+    [texDescriptor setDepth:edgelen];
+    [texDescriptor setStorageMode:MTLStorageModeManaged];
+    [texDescriptor setPixelFormat:MTLPixelFormatRGBA32Float];
+    [texDescriptor setMipmapLevelCount:1];
+    id<MTLTexture> tex = [[device newTextureWithDescriptor:texDescriptor] autorelease];
+    
+    const void* texData = float4AdaptedLutValues.data();
+    
+    [tex replaceRegion:MTLRegionMake3D(0, 0, 0, edgelen, edgelen, edgelen)
+           mipmapLevel:0
+                 slice:0
+             withBytes:texData
```

Comment:
[Question] As `texData` is the internal pointer from `float4AdaptedLutValues` which is a local variable, it will be a dangling pointer when leaving the method scope. Does MTLTexture copy the image data?

Reply:
replaceRegion will copy the image data immediately to the texture. There is no async operation happening here.